implementations/rust/daemon/src/node.rs

```diff
@@ -0,0 +1,160 @@
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::sync::mpsc::{self, Sender};
+use std::thread;
+
+use crate::cli;
+
+use ockam_common::commands::ockam_commands::*;
+use ockam_message::message::{Message as OckamMessage, MessageType, Route};
+use ockam_router::router::Router;
+use ockam_transport::transport::UdpTransport;
+use ockam_vault::{types::*, Vault};
+
+pub struct Node;
+
+impl Node {
+    pub fn new<V: Vault>(mut vault: V, config: Config) -> Sender<Vec<u8>> {
+        // TODO: determine best way to check for existing vault key, and if `1` is always
+        // the identity for ockamd to use from a persisted filesystem vault.
+
+        // let key = vault.secret_export(SecretKeyContext::Memory(1)).unwrap();
+        // println!("key 1: {:?}", key);
+
+        let kex_attrs = SecretKeyAttributes {
+            xtype: SecretKeyType::P256,
+            persistence: SecretPersistenceType::Persistent,
+            purpose: SecretPurposeType::KeyAgreement,
+        };
+        let _kex_secret_ctx = vault
+            .secret_generate(kex_attrs)
+            .expect("failed to create secret for key agreement");
+
+        // create the input and output channel pairs
+        let (tx_in, rx_in) = mpsc::channel();
+
+        let (transport_tx, transport_rx) = mpsc::channel();
+        let (router_tx, router_rx) = mpsc::channel();
+        // let (channel_tx, channel_rx) = mpsc::channel();
+        // let channel_tx_for_node = channel_tx.clone();
+        // let router_tx_for_channel = router_tx.clone();
+
+        let transport_tx_for_node = transport_tx.clone();
+        let mut router = Router::new(router_rx);
+
+        // create the transport, currently UDP-only, using the first configured onward route and
+        // poll it for messages on another thread
+        if config.output_to_stdout {
+            if let Ok(mut transport) = UdpTransport::new(
+                transport_rx,
+                transport_tx,
+                router_tx,
+                config.local_host.to_string().as_str(),
+            ) {
+                thread::spawn(move || {
+                    while transport.poll() && router.poll() {
+                        thread::sleep(std::time::Duration::from_millis(100));
+                    }
+                });
+            }
+        }
+
+        let route_config = config.clone();
+        let remote_addr = route_config
+            .onward_route
+            .expect("invalid address provided for output")
+            .clone();
+        thread::spawn(move || {
+            loop {
+                if let Ok(data) = rx_in.recv() {
```

Comment:
Right now you're ignoring `Err`. What I mean here is shouldn't you handle it with `eprintln` for now? I agree a logger in the future would be better

Reply:
I see what you're saying. Sorry, I was looking at a different `eprintln!` use inside the `if let` block. Yes, I'll print the error, but now that you've brought my attention to it, I wonder if this should actually terminate the program after printing the error.. otherwise I think it will just print errors repeatedly until the process is killed.

Reply:
```rust
eprintln!("fatal error: failed to read from input");
std::os::exit(1);
```

basically, if this CLI stdin reader can't read from it's input channel then it's no use running at all... os::exit(1) is probably the easy way out, but happy to consider alternatives if you have a suggestion.

Reply:
That sounds like a great solution.